winpr/libwinpr/sspi/Negotiate/negotiate.c

```diff
@@ -160,143 +326,544 @@ static BOOL negotiate_get_config(BOOL* kerberos, BOOL* ntlm)
+	SECURITY_STATUS status;
+	ULONG inner_mech_list_len = 0;
+	BYTE* p;
+	const Mech* mech;
 
-	if (!negotiate_get_config(&kerberos, &ntlm))
-		return status;
+	if (!phCredential || !SecIsValidHandle(phCredential))
+		return SEC_E_NO_CREDENTIALS;
 
-	context = (NEGOTIATE_CONTEXT*)sspi_SecureHandleGetLowerPointer(phContext);
+	context = sspi_SecureHandleGetLowerPointer(phContext);
+	creds = sspi_SecureHandleGetLowerPointer(phCredential);
+	if (pInput)
+		input_buffer = sspi_FindSecBuffer(pInput, SECBUFFER_TOKEN);
+	if (pOutput)
+		output_buffer = sspi_FindSecBuffer(pOutput, SECBUFFER_TOKEN);
 
 	if (!context)
 	{
-		context = negotiate_ContextNew();
+		for (int i = 0; i < MECH_COUNT; i++)
+		{
+			if (!creds[i].valid)
+				continue;
+
+			inner_mech_list_len += ASN_TLV_LENGTH(creds[i].mech->oid->length);
 
+			if (init_context.mech) /* We already have an optimistic mechanism */
+				continue;
+
+			/* Use the output buffer to store the optimistic token */
+			CopyMemory(&output_token.mechToken, output_buffer, sizeof(SecBuffer));
+
+			status = MechTable[i].pkg->table_w->InitializeSecurityContextW(
+			    &creds[i].cred, NULL, pszTargetName, fContextReq | creds[i].mech->init_flags,
+			    Reserved1, TargetDataRep, NULL, Reserved2, &init_context.sub_context, &mech_output,
+			    pfContextAttr, ptsExpiry);
+
+			/* If the mechanism failed we can't use it; skip */
+			if (IsSecurityStatusError(status))
+				creds[i].valid = FALSE;
+			else
+				init_context.mech = creds[i].mech;
+		}
+
+		/* No usable mechanisms were found */
+		if (!init_context.mech)
+			return status;
+
+#ifdef WITH_SPNEGO
+		/* If the only available mech is NTLM use it directly otherwise use spnego */
+		if (init_context.mech->oid == &ntlm_OID)
+		{
+			init_context.spnego = FALSE;
+			output_buffer->cbBuffer = output_token.mechToken.cbBuffer;
+			WLog_DBG(TAG, "Using direct NTLM");
+		}
+		else
+		{
+			init_context.spnego = TRUE;
+			init_context.mechTypes.BufferType = SECBUFFER_DATA;
+			init_context.mechTypes.cbBuffer = ASN_TLV_LENGTH(inner_mech_list_len);
+		}
+#else
+		init_context.spnego = FALSE;
+		output_buffer->cbBuffer = output_token.mechToken.cbBuffer;
+#endif
+
+		/* Allocate memory for the new context */
+		context = negotiate_ContextNew(&init_context);
 		if (!context)
-			return SEC_E_INTERNAL_ERROR;
+		{
+			init_context.mech->pkg->table->DeleteSecurityContext(&init_context.sub_context);
+			return SEC_E_INSUFFICIENT_MEMORY;
+		}
 
+		sspi_SecureHandleSetUpperPointer(phNewContext, NEGO_SSP_NAME);
 		sspi_SecureHandleSetLowerPointer(phNewContext, context);
-		sspi_SecureHandleSetUpperPointer(phNewContext, (void*)NEGO_SSP_NAME);
-	}
 
-	/* if Kerberos has previously failed or WITH_GSSAPI is not defined, we use NTLM directly */
-	if (kerberos && !ErrorInitContextKerberos)
+		if (!context->spnego)
+			return status;
+
+		/* Write the SEQUENCE tag */
+		p = negotiate_write_tlv(context->mechTypes.pvBuffer, 0x30, inner_mech_list_len, NULL);
```

Comment:
It's not obvious how the check is done that `context->mechTypes.pvBuffer` has enough space to hold what we have to write in it...

Reply:
It's set in `mechTypes.cbBuffer` and allocated in `negotiate_ContextNew()`